crates/infra/cli/src/toolchains/napi/cli.rs

```diff
@@ -114,3 +119,87 @@ impl NapiCli {
+///
+/// This is necessary to retain extension compatibility with as many systems as possible:
+/// <https://code.visualstudio.com/docs/supporting/requirements#_additional-linux-requirements>.
+fn ensure_correct_glibc_for_vscode(
+    resolver: &NapiResolver,
+    output_dir: &Path,
+    target: &BuildTarget,
+) -> Result<()> {
+    let target_triple = match target {
+        BuildTarget::ReleaseTarget(target) if target.ends_with("-linux-gnu") => target,
+        _ => return Ok(()),
+    };
+
+    let min_glibc = NapiConfig::target_glibc(resolver)?;
+
+    let output_artifact = match target_triple.split('-').next() {
+        Some("x86_64") => "index.linux-x64-gnu.node",
+        Some("aarch64") => "index.linux-arm64-gnu.node",
+        _ => bail!("Unsupported target {target_triple} for `cargo-zigbuild`."),
+    };
+    let output_artifact_path = output_dir.join(output_artifact);
+
+    let is_host_compiling = target_triple.starts_with(std::env::consts::ARCH);
+    if is_host_compiling {
+        let rust_crate_name = resolver.rust_crate_name();
+
+        // Don't clobber the existing output directory.
+        let zigbuild_output = tempfile::tempdir()?;
+
+        // Until `@napi-rs/cli` v3 is released with a fixed `zig` support and a new `--cross-compile`,
+        // we explicitly compile ourselves again with `cargo-zigbuild` to target the desired GLIBC
+        // version, without having to separately compile on the target platform (e.g. via Docker).
+        Command::new("cargo")
+            .arg("zigbuild")
+            .property("-p", rust_crate_name)
+            .flag("--release")
+            .property("--target", format!("{target_triple}.{min_glibc}"))
+            .property("--target-dir", zigbuild_output.path().to_string_lossy())
+            .run()?;
+
+        // Overwrite the existing artifact with the cross-compiled one.
+        let zigbuild_output = zigbuild_output.into_path();
+        let artifact_path = zigbuild_output
+            .join(target_triple)
+            .join("release")
+            .join(format!("lib{rust_crate_name}.so"));
+
+        std::fs::copy(artifact_path, &output_artifact_path)?;
+    } else {
+        // Already cross-compiled with the correct GLIBC version. Just verify for sanity.
+    }
+
+    // Verify that the artifact is compatible with the desired GLIBC version.
+    let library_glibc_version = Command::new("scripts/min_glibc_version.sh")
+        .arg(output_artifact_path.to_string_lossy())
+        .evaluate()?;
+
+    if lenient_semver(&library_glibc_version)? > lenient_semver(&min_glibc)? {
+        bail!("The compiled artifact {output_artifact_path:?} targets GLIBC {library_glibc_version}, which is higher than the minimum specified version {min_glibc}.");
+    }
+
+    Ok(())
+}
+
+/// Like `Version::parse`, but allows for a missing patch version, defaulting to `0`.
+fn lenient_semver(value: &str) -> Result<Version> {
```

Comment:
I suggest using `semver::Comparator` instead, as it supports partial versions, and can also test/compare versions based on a provided operator `>` ...

Reply:
I explicitly wanted to support exactly just optional patch version here, so no operator nor any semver suffixes. If we ended up with `>2.28` or `2.20.0.something` then that's really unexpected in terms of `objdump` and would hint that something is extracted wrongly here, as the printed versions are really simple.

This feature would would be a hidden bug and the logic here is trivial: parse ints separated by `.`, only take 2 or 3, rest is error, so I'll stick with it, if you don't mind.

Reply:
I'll attemt only deserializing the Comparator directly in the package.json, that should help once we move to Rust checks for the glibc version extraction; I'll submit it tomorrow.

Reply:
I could not use the Comparator nor VersionReq - we need to explicitly pass the target GLIBC version to zigbuild and semver does not implement any resolution logic like "pick the minimal version satisfying a given constraint". As such, I've separated the Zig's GLIBC version into a dedicated ADT with custom deserialization and `PartialOrd` impl with `Version` for convenience.

Reply:
I see. If we are only expecting a handful of versions, and they need to be passed in a specific format (`major.minor` without `.patch`), I wonder if we should just have a short list/enum here of explicit strings (`2.18`, `2.19`, ...) we do support, instead of trying to parse any generic `minor` number, then using it to form the correct version.

I think it is a bit confusing to me how this reads from the config, and reviewing a PR that changes:

```json
"glibcVersion": "2.18",
```

To:

```json
"glibcVersion": "2.19",
```

In `package.json` is directly understandable without reading how `glibc.rs` works. And maintaining the handful of versions is trivial, as they won't change much.

However, I wouldn't block on this. Up to you if you see other issues here.

Reply:
I can change the phrasing to include the "target" word - this would make it more clear that we're *targeting* this exact glibc version when building.

Re first point: I actually started with an enum
```rust
enum ZigVersion {
  #[strum(serialize = "2.16")]
  v2_16,
  // ...
}
```

but:
1. the list quickly got out of hand for all of the supported variants (~50 LoC), we also would never use the variants ourselves anyway in code (deserialized)
2. Alternatively, I think there's no point in including only a subset of these values:
  a. I wanted to establish where this list comes from and what it means/how it's used (Zig's supported GLIBC range)
  b. we might as well hardcode the one version we're interested in (I was actually considering it but I don't like hardcoding important configuration constants in code and I'm sure you'd want me to change that)
3. we still need a `Serialize` impl anyway for the json deserialization/validation apart from the strum-derived FromStr

Reply:
Actually, we do use *target* phrasing:
```
  "slangMetadata": {
    "targetGlibc": "2.28"
  },
```

GLIBC is inherently a GNU-only concept (as it's a GNU C Library) and since we're using related terminology like "target", "(target) triples", "glibc" (defines the -gnu environment) immediately next to each other, I think it's very clear that this additionally specifies GLIBC 2.28 requirement for the -gnu targets listed above.

@OmarTawfik Do you have an idea how to better convey this? We could probably also just hardcode the constant, instead and document why it's 2.28 (because it's minimally supported version for VS Code; the code in `glibc.rs` also is very vocal about this particular use case/the reason for it).

Reply:
Valid points! Agreed on the above. I was simply suggesting that just even using a simple string type `targetGlibC: String` and asserting it is one of the supported values (only one value for now is enough. My main worry is readability in the config file, when changing/reviewing diffs there.
Otherwise, I think for how much this will actually change/be reviewed in the foreseable future, we might be over-engineering this :) so I'm happy to go with your intuition here.

Reply:
I think there might be some misunderstanding here, we do exactly that in this PR - we deserialize as string and check if it's `2.16-38`, see `Deserialize` impl. This
``` json
  "slangMetadata": {
    "targetGlibc": "2.28"
  },
```
is how it's already stored in the package.json.

The fact that the Rust struct only has `minor: u8` does not mean it will be deserialized as `u8`; it also has `Display` trait implementation that correctly prints `2.{minor}`. I'm not using more state in the struct (e.g. `String` or two `u8`s) to limit the ability of representing invalid state in the type.

Based on this, I imagine that works as you'd expect and it's readable enough in the package.json 👍